### PR TITLE
Advanced annotation support for SQL databases for name, type, indexes and generated fields

### PR DESCRIPTION
**Pull request by hasathcharu**
## Purpose
To add advanced annotation support for SQL databases for name mapping, type mapping, relation mapping, indexes and auto generated columns to support database introspection. Compiler validations for the new annotations are also added in this PR.

Fixes: 
- [ballerina-platform/ballerina-library/issues/6013](https://github.com/ballerina-platform/ballerina-library/issues/6013)
- [ballerina-platform/ballerina-library/issues/6068](https://github.com/ballerina-platform/ballerina-library/issues/6068)

## Checklist
- [x] Linked to an issue
- [ ] Updated the specification
- [x] Updated the changelog
- [x] Added tests
- [ ] Checked native-image compatibility

### CODE REVIEW

File: ballerina/sql_client.bal [hunk 1]
```diff
@@ -420,11 +419,17 @@ public isolated client class SQLClient {
     }
 
     private isolated function getUpdateQuery(record {} updateRecord) returns sql:ParameterizedQuery|persist:Error {
-        return sql:queryConcat(`UPDATE `, stringToParameterizedQuery(self.escape(self.tableName)), ` SET `, check self.getSetClauses(updateRecord));
+        if (self.dataSourceSpecifics == MSSQL_SPECIFICS) {
```
Comment (Bhashinee) [inline ballerina/sql_client.bal:422]:
```suggestion
        if self.dataSourceSpecifics == MSSQL_SPECIFICS {
```

File: ballerina/sql_client.bal [hunk 2]
```diff
@@ -420,11 +419,17 @@ public isolated client class SQLClient {
...
+            return sql:queryConcat(`UPDATE `, stringToParameterizedQuery(self.escape(self.entityName)), ` SET `, check self.getSetClauses(updateRecord), ` FROM `, stringToParameterizedQuery(self.escape(self.tableName)), ` `, stringToParameterizedQuery(self.escape(self.entityName)));
+        }
+        return sql:queryConcat(`UPDATE `, stringToParameterizedQuery(self.escape(self.tableName)), ` AS `, stringToParameterizedQuery(self.escape(self.entityName)), ` SET `, check self.getSetClauses(updateRecord));
     }
 
     private isolated function getDeleteQuery() returns sql:ParameterizedQuery {
-        return sql:queryConcat(`DELETE FROM `, stringToParameterizedQuery(self.escape(self.tableName)));
+        if (self.dataSourceSpecifics == MSSQL_SPECIFICS) {
```
Comment (Bhashinee) [inline ballerina/sql_client.bal:429]:
```suggestion
        if self.dataSourceSpecifics == MSSQL_SPECIFICS {
```

File: ballerina/tests/init-tests.bal [hunk 3]
```diff
@@ -53,26 +53,37 @@ configurable record {|
...
 
     //MSSQL
-    mssql:Client mssqlDbClient = check new (host = mssql.host, user = mssql.user, password = mssql.password, port = mssql.port);
+    check MSSQLTests();
+
+    // PostgreSQL
+    check PostgreSQLTests();
+}
+
+function MySQLTests() returns error? {
```
Comment (Bhashinee) [inline ballerina/tests/init-tests.bal:65]: Shall we use the camel case for function names?
```suggestion
function mySqlTests() returns error? {
```

Comment (Bhashinee) [reply ballerina/tests/init-tests.bal:65]: Please check other test function names as well. 
PostgreSQLTests() -> postgreSqlTests()
MSSQLTests() -> msSqlTests()

File: compiler-plugin-test/build.gradle [hunk 4]
```diff
@@ -42,6 +42,8 @@ dependencies {
     implementation group: 'org.testng', name: 'testng', version: "${testngVersion}"
     testImplementation 'org.testng:testng'
     testImplementation group: 'com.google.code.gson', name: 'gson', version: "${gsonVersion}"
+    implementation group: 'io.ballerina.stdlib', name: 'persist-native', version: "1.2.1-20240222-125000-4a9d764"
+    implementation group: 'io.ballerina.stdlib', name: 'persist-compiler-plugin', version: "1.2.1-20240222-125000-4a9d764"
```
Comment (Bhashinee) [inline compiler-plugin-test/build.gradle:46]: Shall we have place holders for versions here?

File: compiler-plugin-test/src/test/java/io/ballerina/stdlib/persist/sql/compiler/CompilerPluginTest.java [hunk 5]
```diff
@@ -0,0 +1,419 @@
...
+import static io.ballerina.stdlib.persist.compiler.DiagnosticsCodes.PERSIST_614;
+import static io.ballerina.stdlib.persist.compiler.DiagnosticsCodes.PERSIST_615;
+import static io.ballerina.stdlib.persist.compiler.DiagnosticsCodes.PERSIST_616;
+import static io.ballerina.stdlib.persist.sql.compiler.TestUtils.getEnvironmentBuilder;
+
+/**
+ * Tests persist compiler plugin.
+ */
+public class CompilerPluginTest {
+
```
Comment (Bhashinee) [inline compiler-plugin-test/src/test/java/io/ballerina/stdlib/persist/sql/compiler/CompilerPluginTest.java:63]:
```suggestion
```

File: compiler-plugin-test/src/test/java/io/ballerina/stdlib/persist/sql/compiler/CompilerPluginTest.java [hunk 6]
```diff
@@ -0,0 +1,419 @@
...
+
+    private Package loadPersistModelFile(String directory, String name) {
+        Path projectDirPath = Paths.get("src", "test", "resources", directory, "persist").
+                toAbsolutePath().resolve(name);
+        SingleFileProject project = SingleFileProject.load(getEnvironmentBuilder(), projectDirPath);
+        return project.currentPackage();
+    }
```
Comment (Bhashinee) [inline compiler-plugin-test/src/test/java/io/ballerina/stdlib/persist/sql/compiler/CompilerPluginTest.java:70]:
```suggestion
    }
    
```

File: compiler-plugin-test/src/test/java/io/ballerina/stdlib/persist/sql/compiler/CompilerPluginTest.java [hunk 7]
```diff
@@ -0,0 +1,419 @@
...
+                },
+                new String[]{
+                        "invalid use of Relation annotation. the field 'car' is an optional type in a 1-1 " +
+                                "relationship. therefore, it cannot have foreign keys."
+                },
+                new String[]{
+                        "(9:4,10:13)"
+                }
+        );
+    }
```
Comment (Bhashinee) [inline compiler-plugin-test/src/test/java/io/ballerina/stdlib/persist/sql/compiler/CompilerPluginTest.java:266]: Let's have a new line before the test functions. It's more readable in that way. Please check other places as well.

File: compiler-plugin-test/src/test/resources/project_1/persist/char.bal [hunk 8]
```diff
@@ -0,0 +1,37 @@
+import ballerina/persist as _;
```
Comment (Bhashinee) [inline compiler-plugin-test/src/test/resources/project_1/persist/char.bal:1]: Shall we add the license header?

File: compiler-plugin-test/src/test/resources/project_1/persist/char.bal [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+import ballerina/time;
+import ballerinax/persist.sql;
+
+public enum AppointmentStatus {
+    SCHEDULED = "SCHEDULED",
+    STARTED = "STARTED",
+    ENDED = "ENDED"
```
Comment (Bhashinee) [inline compiler-plugin-test/src/test/resources/project_1/persist/char.bal:8]:
```suggestion
    SCHEDULED,
    STARTED,
    ENDED
```

File: compiler-plugin-test/src/test/resources/project_1/persist/char.bal [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+}
+
+public enum PatientGender {
+    MALE = "MALE",
+    FEMALE = "FEMALE"
```
Comment (Bhashinee) [inline compiler-plugin-test/src/test/resources/project_1/persist/char.bal:13]: Ditto

File: compiler-plugin-test/src/test/resources/project_1/persist/decimal.bal [hunk 11]
```diff
@@ -0,0 +1,12 @@
...
+
+public type Person record {|
+    readonly int id;
+    string name;
+    @sql:Decimal {precision: [10,2]}
+    int age;
+    @sql:Decimal {precision: [0,2]}
+    decimal salary;
+    decimal height;
+|};
```
Comment (Bhashinee) [inline compiler-plugin-test/src/test/resources/project_1/persist/decimal.bal:12]:
```suggestion
|};

```

File: compiler-plugin-test/src/test/resources/project_1/persist/entity_name_mapping.bal [hunk 12]
```diff
@@ -0,0 +1,56 @@
+import ballerina/persist as _;
```
Comment (Bhashinee) [inline compiler-plugin-test/src/test/resources/project_1/persist/entity_name_mapping.bal:1]: Shall we add the license header?

Comment (Bhashinee) [reply compiler-plugin-test/src/test/resources/project_1/persist/entity_name_mapping.bal:1]: Please check the rest of the files as well.

File: compiler-plugin-test/src/test/resources/project_1/persist/char.bal (hunk 10 above)
Comment (Bhashinee) [reply compiler-plugin-test/src/test/resources/project_1/persist/char.bal:13]: Please check the other places as well.

File: compiler-plugin-test/src/test/resources/project_1/persist/decimal.bal (hunk 11 above)
Comment (Bhashinee) [reply compiler-plugin-test/src/test/resources/project_1/persist/decimal.bal:12]: Please check the other places as well.

File: compiler-plugin/build.gradle [hunk 13]
```diff
@@ -33,7 +33,8 @@ dependencies {
     implementation group: 'org.ballerinalang', name: 'ballerina-parser', version: "${ballerinaLangVersion}"
     implementation group: 'org.ballerinalang', name: 'toml-parser', version: "${ballerinaLangVersion}"
     implementation group: 'org.ballerinalang', name: 'formatter-core', version: "${ballerinaLangVersion}"
-    implementation group: 'io.ballerina.stdlib', name: 'persist-native', version: "${stdlibPersistVersion}"
+    implementation group: 'io.ballerina.stdlib', name: 'persist-native', version: "1.2.1-20240222-125000-4a9d764"
+    implementation group: 'io.ballerina.stdlib', name: 'persist-compiler-plugin', version: "1.2.1-20240222-125000-4a9d764"
```
Comment (Bhashinee) [inline compiler-plugin/build.gradle:37]: Let's have placeholders for these 2

File: compiler-plugin-test/src/test/resources/modelvalidator/Ballerina.toml [hunk 14]
```diff
@@ -0,0 +1,14 @@
...
+
+[[tool.persist]]
+id = "generate-db-client"
+targetModule = "project_1.test"
+options.datastore = "mysql"
+filePath = "persist/model.bal"
+
+[persist]
+datastore = "mysql"
+module = "Demo.db"
```
Comment (daneshk) [inline compiler-plugin-test/src/test/resources/modelvalidator/Ballerina.toml:14]:
```suggestion
module = "Demo.db"

```

File: gradle.properties [hunk 15]
```diff
@@ -34,7 +34,7 @@ stdlibHttpVersion=2.10.0
 stdlibSqlVersion=1.11.0
 
 # Level 09
-stdlibPersistVersion=1.2.1-20231130-100000-cca5690
+stdlibPersistVersion=1.3.0-20240311-170600-561b857
```
Comment (daneshk) [inline gradle.properties:37]:
```suggestion
stdlibPersistVersion=1.3.0-20240311-220000-f31520a
```

File: ballerina/tests/init-tests.bal (hunk 3 above)
Comment (Bhashinee) [reply ballerina/tests/init-tests.bal]: Sorry didn't notice this before. Let's start the function names with a verb. 
ex:- testPostgreSql()

File: compiler-plugin-test/src/test/resources/modelvalidator/persist/compilationerror.bal [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+
+import ballerina/persist as _;
+
+public type Person record {|
+    readonly int nic;
+    string name;
+    int age;
+    string city;
+    any 'type;
+|};
```
Comment (daneshk) [inline compiler-plugin-test/src/test/resources/modelvalidator/persist/compilationerror.bal:25]:
```suggestion
|};

```

**Review by daneshk — APPROVED** [review]

**Pull request merged**